qa/rpc-tests/cashlibtest.py

```diff
@@ -0,0 +1,161 @@
+        # Note for this template I readied 4 nodes but only started 2
+
+        # Now interconnect the nodes
+        connect_nodes_bi(self.nodes, 0, 1)
+        # Let the framework know if the network is fully connected.
+        # If not, the framework assumes this partition: (0,1) and (2,3)
+        # For more complex partitions, you can't use the self.sync* member functions
+        self.is_network_split = False
+        self.sync_all()
+
+    def run_test(self):
+
+        faulted = False
+        try:
+            cashlib.spendscript(OP_1)
+        except AssertionError:
+            faulted = True
+            pass
+        assert faulted, "only data in spend scripts"
+
+        try:
+            cashlib.signTxInput(b"", 0, 5, b"", b"", cashlib.SIGHASH_ALL)
+        except AssertionError:
+            faulted = True
+            pass
+        assert faulted, "not signing with bitcoin cash forkid"
+
+        # grab inputs from 2 different full nodes and sign a single tx that spends them both
+        wallets = [self.nodes[0].listunspent(), self.nodes[1].listunspent()]
+        inputs = [x[0] for x in wallets]
+        privb58 = [self.nodes[0].dumpprivkey(inputs[0]["address"]), self.nodes[1].dumpprivkey(inputs[1]["address"])]
+
+        privkeys = [decodeBase58(x)[1:-5] for x in privb58]
+        pubkeys = [cashlib.pubkey(x) for x in privkeys]
+
+        tx = CTransaction()
+        for i in inputs:
+            tx.vin.append(CTxIn(COutPoint(i["txid"], i["vout"]), b"", 0xffffffff))
+
+        destPrivKey = cashlib.randombytes(32)
+        destPubKey = cashlib.pubkey(destPrivKey)
+        destHash = cashlib.addrbin(destPubKey)
+
+        output = CScript([OP_DUP, OP_HASH160, destHash, OP_EQUALVERIFY, OP_CHECKSIG])
+
+        amt = int(sum([x["amount"] for x in inputs]) * BCH)
+        tx.vout.append(CTxOut(amt, output))
+
+        sighashtype = 0x41
+        n = 0
+        for i, priv in zip(inputs, privkeys):
+            sig = cashlib.signTxInput(tx, n, i["amount"], i["scriptPubKey"], priv, sighashtype)
+            tx.vin[n].scriptSig = cashlib.spendscript(sig)  # P2PK
+            n += 1
+
+        txhex = hexlify(tx.serialize()).decode("utf-8")
+        txid = self.nodes[0].sendrawtransaction(txhex)
+
+        assert txid == hexlify(cashlib.txid(txhex)[::-1]).decode("utf-8")
+
+        # Now spend the created output to an anyone can spend address
+        tx2 = CTransaction()
+        tx2.vin.append(CTxIn(COutPoint(cashlib.txid(txhex), 0), b"", 0xffffffff))
+        tx2.vout.append(CTxOut(amt, CScript([OP_1])))
+        sig2 = cashlib.signTxInput(tx2, 0, amt, output, destPrivKey, sighashtype)
+        tx2.vin[0].scriptSig = cashlib.spendscript(sig2, destPubKey)
+
+        tx2id = self.nodes[0].sendrawtransaction(hexlify(tx2.serialize()).decode("utf-8"))
+
+        # Check that all tx were created, and commit them
+        assert self.nodes[0].getmempoolinfo()["size"] == 2
+        self.nodes[0].generate(1)
+
+
```

Comment:
I would do one last check here to make sure the block was propagated correctly and that the mempool is empty.

+ self.sync_all()
+ assert self.nodes[0].getmempoolinfo()["size"] == 0

Reply:
I did a sync_blocks and check.  We should use sync_all only when absolutely necessary because mempools may not ever get in sync if the bloom filter has a false positive.  This makes the test fail (very rarely, depending on the txid) forcing us to rekick travis.